src/Raven.Server/Documents/Sharding/BucketStatsHolder.cs

```diff
@@ -0,0 +1,119 @@
+            {
+                changeVector = currentMergedCv.MergeWith(changeVector, _ctx);
+            }
+
+            _mergedChangeVectors[bucket] = changeVector;
+        }
+
+        public void UpdateBucketStatsTreeBeforeCommit(Transaction tx)
+        {
+            int bucketStatsSize = sizeof(Documents.BucketStats);
+            var tree = tx.ReadTree(ShardedDocumentsStorage.BucketStatsSlice);
+
+            foreach ((int bucket, Documents.BucketStats inMemoryStats) in _values)
+            {
+                using (tx.Allocator.Allocate(sizeof(int), out var keyBuffer))
+                {
+                    *(int*)keyBuffer.Ptr = bucket;
+                    var keySlice = new Slice(keyBuffer);
+                    var readResult = tree.Read(keySlice);
+
+                    Documents.BucketStats stats;
+                    IDisposable cvScope = null;
+                    Slice cvSlice = default;
+
+                    if (readResult == null)
+                    {
+                        stats = inMemoryStats;
+                    }
+                    else
+                    {
+                        stats = *(Documents.BucketStats*)readResult.Reader.Base;
+                        stats.Size += inMemoryStats.Size;
+                        stats.NumberOfDocuments += inMemoryStats.NumberOfDocuments;
+                        stats.LastModifiedTicks = inMemoryStats.LastModifiedTicks;
+
+                        cvScope = stats.GetMergedChangeVector(tx.Allocator, readResult.Reader, out cvSlice);
+                    }
+
+                    if (stats.Size == 0 && stats.NumberOfDocuments == 0)
+                    {
+                        tree.Delete(keySlice);
+                        continue;
+                    }
+
+                    if (_mergedChangeVectors.TryGetValue(bucket, out var changeVector))
+                    {
+                        using (cvScope)
+                        {
+                            changeVector = changeVector.MergeWith(cvSlice.HasValue ? cvSlice.ToString() : null, _ctx);
+                            cvScope = Slice.From(tx.Allocator, changeVector, out cvSlice);
+                        }
+                    }
+
+                    using (cvScope)
+                    using (tree.DirectAdd(keySlice, bucketStatsSize + cvSlice.Size, out var ptr))
+                    {
+                        *(Documents.BucketStats*)ptr = stats;
+                        if (cvSlice.HasValue)
+                            cvSlice.CopyTo(ptr + bucketStatsSize);
+                    }
+                }
+            }
+
+            _ctx?.Dispose();
+            _ctx = null;
+            _values.Clear();
+            _mergedChangeVectors.Clear();
+        }
+
+        private ChangeVector TableValueToChangeVector(int changeVectorIndex, ref TableValueReader value)
+        {
+            if (_ctx == null)
+                _storage.ContextPool.AllocateOperationContext(out _ctx);
```

Comment:
what happened if this tx allocation a context, but got an exception somewhere in the middle (like Concurrency exception..)
This context would be somehow disposed?

Reply:
fixed.

changed `Transaction.Owner` to point to the `DocumentOperationContext` that created the transaction, instead of pointing to the `DocumentDatabase`

so now we're using this context instead of allocating a new one
and we can still access the database when needed via `docsContext.DocumentDatabase`